### ISSUE TITLE
Getting Sumatra to work with Gvim on Windows 10 machines

### ISSUE THREAD

**Issue by llinfeng** [lervag/vimtex#1585]
In this post, I wish to address all setup problems with Sumatra + `gvim.exe` on Windows 10 machines. The installation step is omitted.

# Principles:
1. If your `gvim.exe` is `32-bit`, you also want to use `32-bit` Sumatra viewer.
   [**Source1**](https://github.com/lervag/vimtex/issues/1345#issuecomment-476375406)
2. Make sure `SumatraPDF` is "callable" through the "Run-window" (Win + R), by adding the absolute path of the `32-bit` Sumatra viewer to your the `Path` environmental variable.
   [**Source2**](https://github.com/lervag/vimtex/issues/416#issue-145544363)

## `.vimrc` settings

```vim
let g:vimtex_view_general_viewer = 'SumatraPDF'
let g:vimtex_view_general_options
    \ = '-reuse-instance -forward-search @tex @line @pdf'
let g:vimtex_view_general_options_latexmk = '-reuse-instance'
```

## Debugging checklist
Based on _personal experience_, forward/backward search may or may not work out of the box. Here are the scenarios that I solved.

### What if `:VimtexView` is not callable?
Well, check if you have set the following line in your `.vimrc`:
```
let g:vimtex_view_enabled = 0
```
With this option set as `0`, `:VimtexView` won't be assigned at all.

### What if `:VimtexView` does nothing?
1. Check if you are having a `32-bit` Vim but a `64-bit` SumatraPDF.exe;
    * If so, remove the `64-bit` SumatraPDF.exe and install the `32-bit` version from [Sumatra PDF download page](https://www.sumatrapdfreader.org/downloadafter.html)
2. Execute `SumatraPDF` in the "Run-window" (Win+R), and check its version.
    * In rare cases, there can be multiple directories in `Path` that contains an executable named as `SumatraPDF.exe`. You'll need to hunt down the path with the wrong version of `SumatraPDF.exe` and remove it (or, put it towards the end among the elements in `Path`).

### What if backward search/jump does not work?
Try to put the following setting to Sumatra's Options window
```
gvim --servername GVIM --remote-send "<C-\><C-n>:drop %f<CR>:%l<CR>:normal! zzzv<CR>:execute 'drop ' . fnameescape('%f')<CR>:%l<CR>:normal! zzzv<CR>:call remote_foreground('GVIM')<CR><CR>"
```
![image](https://user-images.githubusercontent.com/5890407/73373047-25799f80-4286-11ea-973a-8b351854cb99.png)

## Also, add `SumatraPDF` to Path
![image](https://user-images.githubusercontent.com/5890407/73371077-2a891f80-4283-11ea-9073-4bd97d59b649.png)

**Comment by lervag** [comment]
Thank you for this. It seems useful. Perhaps you could suggest a version for including in the vimtex docs in the faq section? And/Or the [wiki sections](https://github.com/lervag/vimtex/wiki)?

Note: I took the liberty of editing your post by updating the `.vimrc` settings. You don't need to set `g:vimtex_view_enabled` (it is enabled by default), and you don't need to specify `let g:vimtex_view_method = 'general'` (also default).

**Comment by lervag** [comment]
Also, if you meant to suggest any actions for me specifically, I missed them. So I'm closing this issue now, but feel free to reopen if I misunderstood anything.

**Comment by llinfeng** [comment]
@lervag I meant to document all the loopholes that took me hours to figure out. I totally agree with your edits to the "necessary setting for Sumatra to work with `vimtex`".

**Comment by lervag** [comment]
I'm very happy you've shared this. It would be even better if you suggested a PR to add some of this to the docs. But I'll try to remember this thread if anyone should raise related issues.

**Comment by bratekarate** [comment]
Thanks a lot, works great for me!

Just a minor addition: It's possible to configure the inverse search per samutra option file or with a CLI flag. Makes it easier to reuse the setup on multiple machines.

`%LOCALAPPDATA%\SumatraPDF\SumatraPDF-settings.txt` (that's where I found my config):
```
InverseSearchCmdLine = gvim --servername GVIMTEX --remote-send "<C-\><C-n>:drop %f<CR>:%l<CR>:normal! zzzv<CR>:execute 'drop ' . fnameescape('%f')<CR>:%l<CR>:normal! zzzv<CR>:call remote_foreground('GVIMTEX')<CR><CR>"
```
I actually rather intended to use `g:vimtex_view_general_options` to provide the `-inverse-search` flag, but it didn't work for me. Maybe the search command is too complex or I didn't escape it right.

Just to show another variant, I use servername `GVIMTEX` to control which vim instance is used. Then it's possible to create a shortcut named something like `gVimtex` with something like the following target:

```
<PATH_TO_VIM>\gvim.exe --servername GVIMTEX %*
```
You can add the containing directory to PATH and open files like `gvim.lnk FILE`. I haven't figured out yet how to configure that `.tex` files are always opened with this command on double click. But it's probably possible as well.

**Comment by llinfeng** [comment]
One note to add: if `SyncTeX` fails to work for some reason, the forward/backward search may not work for certain included files. In one edge-case where [`todonotes` breaks SyncTex](https://www.overleaf.com/learn/how-to/SyncTeX_Errors), things are harder to troubleshoot: for certain included files, forward/backward search works just fine. Yet, for another set of files, well, forward/backward search won't work.

For this particular, case, a simple workaround [exists](https://github.com/henrikmidtiby/todonotes/issues/51#issuecomment-890404454). Of course, the fix is to update `todonotes` to version 1.1.4 or higher.
